src/slate-editor/slate-utils.ts

```diff
@@ -1,81 +1,143 @@
+//   return elt?.getBoundingClientRect();
+// }
+
+// FIXME: this seems unused... at least by CLUE.
+// export function getContentHeight(editor: Editor) {
+//   const { document } = editor.value;
+//   const firstBounds = getBoundingRectForBlock(editor, document.nodes.first());
+//   const lastBounds = getBoundingRectForBlock(editor, document.nodes.last());
+//   return firstBounds && lastBounds
+//           ? lastBounds.bottom - firstBounds.top
+//           : undefined;
+// }
+
+export const toggleBlock = (editor: Editor, format: string) => {
+  const isActive = isBlockActive(
+    editor,
+    format,
+    TEXT_ALIGN_TYPES.includes(format) ? 'align' : 'type'
+  );
+  const isList = LIST_TYPES.includes(format);
+
+  Transforms.unwrapNodes(editor, {
+    match: n =>
+      !Editor.isEditor(n) &&
+      SlateElement.isElement(n) &&
+      LIST_TYPES.includes(n.type) &&
+      !TEXT_ALIGN_TYPES.includes(format),
+    split: true,
   });
-  return markIndex;
-}
+  let newProperties; // FIXME: add back type
+  if (TEXT_ALIGN_TYPES.includes(format)) {
+    newProperties = {
+      align: isActive ? undefined : format,
+    };
+  } else {
+    newProperties = {
+      type: isActive ? 'paragraph' : isList ? 'list-item' : format,
+    } as any;
+  }
+  Transforms.setNodes<SlateElement>(editor, newProperties);
 
-export function findActiveMark(value: Value, format: EFormat) {
-  return value.activeMarks.find(function(mark) { return mark?.type === format; });
-}
+  if (!isActive && isList) {
+    const block = { type: format, children: [] } as any;
+    Transforms.wrapNodes(editor, block);
+  }
+};
 
-export function hasActiveMark(value: Value, format: EFormat) {
-  return !!findActiveMark(value, format);
-}
+export const isBlockActive = (editor: Editor, format: string, blockType = 'type') => {
+  const { selection } = editor;
+  if (!selection) return false;
 
-export function isBlockOfType(node: Node | undefined, format: EFormat) {
-  return Block.isBlock(node) && (node.type === format);
-}
+  const [match] = Array.from(
+    Editor.nodes(editor, {
+      at: Editor.unhangRange(editor, selection),
+      match: n =>
+        !Editor.isEditor(n) &&
+        SlateElement.isElement(n) &&
+        (n as any)[blockType] === format,
+    })
+  );
 
-export function isInlineOfType(node: Node | undefined, format: EFormat) {
-  return Inline.isInline(node) && (node.type === format);
-}
+  return !!match;
+};
 
-// returns whether there is a selected block of the specified type.
-// value.blocks represents the closest selected block(s), i.e. it may
-// indicate that a list item is selected without indicating its list.
-export function hasBlock(value: Value, format: EFormat) {
-  return value.blocks.some(node => isBlockOfType(node, format));
-}
+export const isMarkActive = (editor: Editor, format: string) => {
+  const marks = Editor.marks(editor) as CustomMarks;
+  return !!marks?.[format as MarkType];
+};
```

Comment:
@kswenson I might be reading this wrong, but I don't think this will always do what is expected. The slate code has a comment above `Editor.marks(editor)` of "Get the marks that would be added to text at the current selection."

This is similar to how the docs talk about marks: https://docs.slatejs.org/concepts/07-editor#:~:text=the%20event%20loop.)-,the%20marks%20property,-stores%20formatting%20to 

In both cases these are the marks that would be added to text that is inserted at the current location. Looking at the `Editor.marks(editor)`  code if the Editor.marks property is null then this will do what I think is expected: get the list of all marks that are applied to the current selection.

But if Editor.marks is not null, then its value will be returned instead. As I understand it Editor.addMark will set Editor.marks if the current selection is collapsed. And slate-editor has `toggleMark` which calls this.  Maybe this all works the way we want it but it is certainly complex.

I started digging into this when I read that marks were removed in 0.50 so then I was confused about why slate-editor is still referring to marks. 

I also don't understand how `Editor.marks` and `Editor.marks(editor)` can co-exist. It seems `marks` is both a function and an array.

Reply:
I've looked into this, and here's what's going on as I understand it.
- `marks` was removed from `text` in v0.50, but the concept still exists, and the marks are now direct members of `Text` rather than indirect (it's now `Text.bold` instead of `Text.marks.bold`).
- `editor.marks` are the marks set by the user, which will apply the next time the user types. These can be specified even if no text currently has the selected marks. I think this is the array you're talking about above.
- `Editor.marks(editor)` is a static type function (I'm still not sure of the exact status in javascript, but it seems like a static class function). It does something a little more complex, returning the user specified `editor.marks` if they exist, or else finding the `marks` of the currently selected text and returning those. I believe the intention of this function is to return whatever `marks` will apply to the text that the user writes, wherever those `marks` come from.

If I'm right about the above, I think this is a reasonable way to determine which buttons in the toolbar to make active. I'm not sure how else this function is being used.

Reply:
I now see that Tejal called this out as a bug: https://www.pivotaltracker.com/story/show/184102491

Is the desired behavior here to apply all marks from selected text to the text you replace it with? Or is it just to highlight the buttons of marks used by any selected text (in which case what marks should be applied when the user types and replaces the selected text)? The current behavior is to apply the marks of the first selected text to the replacement text, and so those marks have their buttons highlighted.

Reply:
In the immediate term the goal is to do whatever's easiest, which is probably to accept the current behavior. In the longer term I think the goal should be to conform to whatever the standard practice is for rich text editors, which I _think_ is to only highlight properties that are consistent across all selected text, i.e. not to highlight buttons when only some of the text matches the format. This makes the most sense to me as a user because then clicking the button will first apply the format to all of the text, which will then highlight the button, so that if clicked again it will remove the highlight.